main.rb

```diff
@@ -0,0 +1,95 @@
+module Enumerable
+  def my_each
+    return to_enum unless block_given?
+
+    i = 0
+    arr = to_a
+    while i < arr.length
+      yield(arr[i])
+      i += 1
+    end
+    self
+  end
+  [1, 2, 3].my_each { |n| puts "#{n}!" }
+
+  def my_each_each_with_index
+    return to_enum unless block_given?
+
+    i = 0
+    arr = to_a
+    while i < arr.length
+      yield(arr[i], i)
+      i += 1
+    end
+    self
+  end
+  [1, 2, 3].my_each_each_with_index { |_n, x| print x }
+
+  def my_select
+    return to_enum unless block_given?
+
+    arr = to_a
+    final_array = []
+    arr.my_each { |x| final_array.push[x] if yield(x) }
+  end
+  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10].my_select { |x| puts x if x.even? }
+
+  def my_all?
+    if block_given?
+      my_each { |x| return true if yield(x) == true }
+    else
+      my_each { |x| return true if x == true }
+    end
+    true
+  end
+
+  def my_any?
+    if block_given?
+      my_each { |x| return false unless yield(x) == true }
+    else
+      my_each { |x| return false unless x == true }
+    end
+    false
+  end
+
+  def my_none?
+    return to_enum unless block_given?
+
+    !my_any?
+  end
+
+  def my_count(_args = nil)
+    x = 0
+    if block_given?
+      my_each { |z| x += 1 if yield(z) == true }
+    elsif num.nil?
+      c = length
+    else
+      my_each { |z| x += 1 if z == num }
+    end
+    c
+  end
+
+  def my_map(args)
+    final = []
+    if args == true
+      my_each do |x|
+        final.push(args).call(x)
+      end
+
+      final
+    else
+      self
+    end
+  end
```

Comment:
```rb
2.6.5 :096 > array
 => [0, 8, 5, 2, 3, 3, 2, 7, 5, 5, 1, 4, 6, 6, 4, 4, 5, 4, 7, 6, 0, 7, 6, 5, 8, 2, 0, 2, 8, 3, 6, 7, 4, 4, 1, 4, 7, 3, 7, 1, 1, 7, 3, 4, 2, 4, 6, 5, 2, 6, 1, 3, 6, 7, 4, 1, 6, 2, 1, 1, 8, 3, 2, 5, 4, 7, 8, 5, 3, 2, 7, 0, 2, 1, 3, 2, 5, 1, 1, 3, 8, 7, 1, 0, 3, 7, 6, 8, 6, 3, 5, 6, 5, 1, 3, 0, 8, 2, 8, 4] 
2.6.5 :097 > block = proc { |num| num < 4 }
2.6.5 :098 > array.my_map { |num| num < 10 }
Traceback (most recent call last):
        7: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `<main>'
        6: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `eval'
        5: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `<main>'
        4: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `load'
        3: from /home/maelfosso/.rvm/gems/ruby-2.6.5/gems/irb-1.2.4/exe/irb:11:in `<top (required)>'
        2: from (irb):98
        1: from main.rb:73:in `my_map'
ArgumentError (wrong number of arguments (given 0, expected 1))
2.6.5 :099 > array.map { |num| num < 10 }
 => [true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true] 
2.6.5 :100 > 
```
`#my_map` must return a new array with the results of running a given block once for every element in enum

Reply:
```rb
2.6.5 :102 > array
 => [0, 8, 5, 2, 3, 3, 2, 7, 5, 5, 1, 4, 6, 6, 4, 4, 5, 4, 7, 6, 0, 7, 6, 5, 8, 2, 0, 2, 8, 3, 6, 7, 4, 4, 1, 4, 7, 3, 7, 1, 1, 7, 3, 4, 2, 4, 6, 5, 2, 6, 1, 3, 6, 7, 4, 1, 6, 2, 1, 1, 8, 3, 2, 5, 4, 7, 8, 5, 3, 2, 7, 0, 2, 1, 3, 2, 5, 1, 1, 3, 8, 7, 1, 0, 3, 7, 6, 8, 6, 3, 5, 6, 5, 1, 3, 0, 8, 2, 8, 4] 
2.6.5 :103 > block = proc { |num| num < 4 }
2.6.5 :104 > array.my_map(&block)
Traceback (most recent call last):
        7: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `<main>'
        6: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `eval'
        5: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `<main>'
        4: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `load'
        3: from /home/maelfosso/.rvm/gems/ruby-2.6.5/gems/irb-1.2.4/exe/irb:11:in `<top (required)>'
        2: from (irb):104
        1: from main.rb:73:in `my_map'
ArgumentError (wrong number of arguments (given 0, expected 1))
2.6.5 :105 > array.map(&block)
 => [true, false, false, true, true, true, true, false, false, false, true, false, false, false, false, false, false, false, false, false, true, false, false, false, false, true, true, true, false, true, false, false, false, false, true, false, false, true, false, true, true, false, true, false, true, false, false, false, true, false, true, true, false, false, false, true, false, true, true, true, false, true, true, false, false, false, false, false, true, true, false, true, true, true, true, true, false, true, true, true, false, false, true, true, true, false, false, false, false, true, false, false, false, true, true, true, false, true, false, false] 
2.6.5 :106 > 
```

Reply:
```rb
2.6.5 :108 > array
 => [0, 8, 5, 2, 3, 3, 2, 7, 5, 5, 1, 4, 6, 6, 4, 4, 5, 4, 7, 6, 0, 7, 6, 5, 8, 2, 0, 2, 8, 3, 6, 7, 4, 4, 1, 4, 7, 3, 7, 1, 1, 7, 3, 4, 2, 4, 6, 5, 2, 6, 1, 3, 6, 7, 4, 1, 6, 2, 1, 1, 8, 3, 2, 5, 4, 7, 8, 5, 3, 2, 7, 0, 2, 1, 3, 2, 5, 1, 1, 3, 8, 7, 1, 0, 3, 7, 6, 8, 6, 3, 5, 6, 5, 1, 3, 0, 8, 2, 8, 4] 
2.6.5 :109 > array.map.is_a?(Enumerator)
 => true 
2.6.5 :110 > array.my_map.is_a?(Enumerator)
Traceback (most recent call last):
        7: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `<main>'
        6: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `eval'
        5: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `<main>'
        4: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `load'
        3: from /home/maelfosso/.rvm/gems/ruby-2.6.5/gems/irb-1.2.4/exe/irb:11:in `<top (required)>'
        2: from (irb):110
        1: from main.rb:73:in `my_map'
ArgumentError (wrong number of arguments (given 0, expected 1))
2.6.5 :111 >
```

Reply:
```rb
2.6.5 :113 > my_proc = proc { |num| num > 10 }
2.6.5 :114 > array.my_map(&my_proc)
Traceback (most recent call last):
        7: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `<main>'
        6: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `eval'
        5: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `<main>'
        4: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `load'
        3: from /home/maelfosso/.rvm/gems/ruby-2.6.5/gems/irb-1.2.4/exe/irb:11:in `<top (required)>'
        2: from (irb):114
        1: from main.rb:73:in `my_map'
ArgumentError (wrong number of arguments (given 0, expected 1))
2.6.5 :115 > array.map(&my_proc)
 => [false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false] 
2.6.5 :116 > 
```
`#my_map` must accept a proc as an argument and returns a new array with the results of running the proc once for every element in enum